utils/s2n_asn1_time.c

```diff
@@ -0,0 +1,236 @@
+                current_pos++;
+                state = ON_SUBSECOND;
+                break;
+            case ON_SUBSECOND:
+                if (current_char == '.' || isdigit(current_char)) {
+                    current_pos++;
+                } else {
+                    state = ON_TIMEZONE;
+                }
+                break;
+            case ON_TIMEZONE:
+                if (current_char == 'Z' || current_char == 'z') {
+                    local_time_assumed = 0;
+                    state = FINISHED;
+                } else if (current_char == '-') {
+                    local_time_assumed = 0;
+                    offset_negative = 1;
+                    state = ON_OFFSET_HOURS_DIGIT_1;
+                } else if (current_char == '+') {
+                    local_time_assumed = 0;
+                    offset_negative = 0;
+                    state = ON_OFFSET_HOURS_DIGIT_1;
+                }
+
+                current_pos++;
+                break;
+            case ON_OFFSET_HOURS_DIGIT_1:
+                offset_hours = current_char - '0';
+                current_pos++;
+                state = ON_OFFSET_HOURS_DIGIT_2;
+                break;
+            case ON_OFFSET_HOURS_DIGIT_2:
+                offset_hours = offset_hours * 10 + (current_char - '0');
+                current_pos++;
+                state = ON_OFFSET_MINUTES_DIGIT_1;
+                break;
+            case ON_OFFSET_MINUTES_DIGIT_1:
+                offset_minutes = current_char - '0';
+                current_pos++;
+                state = ON_OFFSET_MINUTES_DIGIT_2;
+                break;
+            case ON_OFFSET_MINUTES_DIGIT_2:
+                offset_minutes = offset_minutes * 10 + (current_char - '0');
+                current_pos++;
+                state = FINISHED;
+                break;
+            default:
+                state = PARSE_ERROR;
+                break;
+        }
+    }
+
+    if (state > ON_TIMEZONE && state < PARSE_ERROR) {
+        time_t clock_data = 0;
+        //ASN1 + and - is in format HHMM. We need to convert it to seconds for the adjustment
+        long gmt_offset = (offset_hours * 3600) + (offset_minutes * 60);
+
+        //go left to go right (it's a little backwards). Negative means that we need to ADD the offset
+        //positive means we need to subtract the offset.
+        if (offset_negative) {
+            gmt_offset = 0 - gmt_offset;
+        }
+
+        //if we detected UTC is being used (please always use UTC), we need to add the detected timezone on the local
+        //machine back to the offset.
+        if (!local_time_assumed) {
+            gmt_offset -= gmt_offset_current;
+        }
+
+        clock_data = mktime(&time);
+
+        //convert to nanoseconds and add the timezone offset.
+        if (clock_data > 0) {
+            *ticks = ((uint64_t) clock_data - gmt_offset) * 1000000000;
+            return 0;
+        }
+    }
+
+    return -1;
+}
```

Comment:
Note for myself: need to really sit down and trace this state machine.

Reply:
its a very simple state machine. It’s probably what you’d get with compiled regex using match groups in C++. One char at a time, Every character in the date gets a state.  

Not sure why I used the word “machine” in that comment though, will fix.

Reply:
oh nevermind machine makes sense when reading the line above it